src/main/java/qupath/bioimageio/spec/Model.java

```diff
@@ -0,0 +1,600 @@
+import java.util.stream.Stream;
+
+
+/**
+ * Resource based on the main model spec.
+ * This extends {@link Resource} to provide more information
+ * relevant for to machine learning models.
+ */
+public class Model extends Resource {
+    private static final Logger logger = LoggerFactory.getLogger(Model.class);
+
+    public void setBaseURI(URI baseURI) {
+        this.baseURI = baseURI;
+    }
+
+    public void setUri(URI uri) {
+        this.uri = uri;
+    }
+
+    URI baseURI;
+    URI uri;
+
+    List<InputTensor> inputs;
+
+    @SerializedName("test_inputs")
+    List<String> testInputs;
+    @SerializedName("test_outputs")
+    List<String> testOutputs;
+
+    // Should be in ISO 8601 format - but preserve string as it is found
+    String timestamp;
+
+    Weights.WeightsMap weights;
+
+    Map<?, ?> config;
+
+    List<OutputTensor> outputs;
+
+    @SerializedName("packaged_by")
+    List<Author> packagedBy;
+
+    ModelParent parent;
+
+    // TODO: Handle run_mode properly
+    @SerializedName("run_mode")
+    Map<?, ?> runMode;
+
+    @SerializedName("sample_inputs")
+    List<String> sampleInputs;
+    @SerializedName("sample_outputs")
+    List<String> sampleOutputs;
+
+
+    /**
+     * Parse a model from a file or directory.
+     * This can either be a yaml file, or a directory that contains a yaml file representing the model.
+     * @param file The file containing the YAML, or its parent directory.
+     * @return The parsed model.
+     * @throws IOException if the model cannot be found or parsed
+     */
+    public static Model parseModel(File file) throws IOException {
+        return parseModel(file.toPath());
+    }
+
+    /**
+     * Parse a model from a path.
+     * This can either represent a yaml file, or a directory that contains a yaml file representing the model.
+     * @param path The path to the file containing the YAML, or its parent directory.
+     * @return The parsed model.
+     * @throws IOException if the model cannot be found or parsed
+     */
+    public static Model parseModel(Path path) throws IOException {
+
+        var pathYaml = findModelRdf(path);
+        if (pathYaml == null) {
+            throw new IOException("Can't find rdf.yaml from " + path);
+        }
+
+        try (var stream = Files.newInputStream(pathYaml)) {
+            var model = parseModel(stream);
+            if (model != null) {
+                model.setBaseURI(pathYaml.getParent().toUri());
+                model.setUri(pathYaml.toUri());
+            }
+            return model;
+        }
+    }
+
+    /**
+     * Parse a model from an input stream.
+     * Note that {@link Model#getBaseURI()} will return null in this case, because the base URI
+     * is unknown.
+     * @param stream A stream of YAML.
+     * @return The parsed model.
+     * @throws IOException if the model cannot be found or parsed
+     */
+    public static Model parseModel(InputStream stream) throws IOException {
```

Comment:
@petebankhead - the other side of the equation lying here, where the main setup for deserializing models (and specifying files etc) is handled